Software Update Architecture: enhancements.

igor-stoppa opened the pull request:
Add clarifications about the rationale for using swupd and
a link to the ClearLinux documentation.

Signed-off-by: Igor Stoppa <igor.stoppa@intel.com>

igor-stoppa commented:
@gvancuts i have added part of the content from the  ELC presentation, so hopefully now it is more clear. Wrt linking, if you are plannign to have the slides on the ostro project site, I'll wait for the url to be available, otherwise I cna add the url from the linux foundation site.
Btw, in a few weeks there should be also a video on youtube, from the presentation at the conference.
There is at least one more presentation that could be added: the one @ipuustin did about security in Ostro.

igor-stoppa commented:
@ipuustin @ereshetova @pohly @rojkov @dbkinder please review, thanks

dbkinder commented:
[skip ci]

rojkov commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
 It can also restore modified files to their pristine state as then were in the
 initial sw image.
 
+Note:
```
nitpick: there's a special directive for notes in rst

```
.. note::
   The version of swupd client...
```

rojkov commented:
the content looks good to me +1

gvancuts commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+  It might list files that are used also by other bundles, but that is by design.
+  Such overlapping is handled by the  Server/Client pair of Software Update tools.
+
+- As long as they come from the same swupdate stream, bundles do not risk to cause
```
s/swupdate/SW update?

gvancuts commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+  The person/entity managing the swupd stream is responsible of vetting the rootfs
+  used during the creation of bundles.
+
+- Bundle creation implies that there is a starting rootfs that provides al the files
```
s/al/all

gvancuts commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+- Bundle creation implies that there is a starting rootfs that provides al the files
+  the bundles will include. How this rootfs is created is irrelevant from the
+  persective of the Swupd Server tool, which creates the bundles.
+  In Ostro, the rootfs is populated through the execution of bitbakre recipes, but
```
s/bitbakre/bitbake

gvancuts commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+  persective of the Swupd Server tool, which creates the bundles.
+  In Ostro, the rootfs is populated through the execution of bitbakre recipes, but
+  in Clear Linux this is done by installing .rpm packages.
+  Compared to pacakges, bundles are a competing way to deliver files to the target
```
s/pacakges/packages

gvancuts commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+  in Clear Linux this is done by installing .rpm packages.
+  Compared to pacakges, bundles are a competing way to deliver files to the target
+  device, however they can be inserted in an existing workflow that is based on
+  pacakges, with minimal modifications.
```
s/pacakges/packages

gvancuts commented:
This is great new version @igor-stoppa ! The only comments left I have are typos. I'm fine leaving them in for now if @dbkinder would do an editorial pass on the doc anyway. Thanks!

igor-stoppa commented:
Fixed all the issues reported, plus added the first part of the howto.

gvancuts commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+
+[Pending: will come in a separate PR]
+
+.. _`swupd layer`: http://git.yoctoproject.org/cgit/cgit.cgi/meta-swupd/tree/docs/GUIDE.md
```
Not sure when but the file name was changed from GUIDE.md to Guide.md so we need to update this link to http://git.yoctoproject.org/cgit/cgit.cgi/meta-swupd/tree/docs/Guide.md

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
-a bare-bones working system.
+Ostro users coming from popular Linux distros are probably already familiar with
+the concept of package.
+How is a bundle different from a pacakge?
```
s/pacakge/package/

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+
+- Bundle creation implies that there is a starting rootfs that provides all the files
+  the bundles will include. How this rootfs is created is irrelevant from the
+  persective of the Swupd Server tool, which creates the bundles.
```
s/persective/perspective/

dbkinder replied:
perspective of the SW update Server tool creating the bundles.

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
 initial sw image.
 
+.. note::
+The version of swupd client that ships with Ostro 1.0 does not support yet
```
almost right... all the "note" text must be indented under the "n" in note:

.. note::
   The version of swupd client...
   signature verification...

The "note" ends with a blank line and text that's no longer indented.

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -17,15 +17,46 @@ Ostro OS based software can be deployed to a target device in two ways:
+In detail, the Software Update functionality is comprised of a pair of
+tools:
+
+ - Software Update Server
```
looks a bit better if you don't indent these bullet items (bring these two bullet lines back to to the left margin)

igor-stoppa replied:
Yes, I got the syntax mixed up with the one from stackexchange :-/

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -17,15 +17,46 @@ Ostro OS based software can be deployed to a target device in two ways:
+
+ - Software Update Server
+
+   A back end tool which generates the bundle/update data stream.
```
reads better if you write out "bundle/update stream" and "consumes/applies the date" as "bundle update stream" (is the "/" really needed?)  and "consumes and applies the data" to clarify the meaning.

igor-stoppa replied:
Bundles are not the same as the update stream: maybe I'm being too literal, but bundles are something one can add/remove without performing any update. I would consider an update to be the replacement of something already installed with a new version of itself.

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -17,15 +17,46 @@ Ostro OS based software can be deployed to a target device in two ways:
-The software update server runs for example,
-on a CI (Continuous Integration) server and not on the device itself.
+The software update server is a tool used to produce the stream of updates
+that, later on, a device can consume, to upgrade itself.
```
that a device can use to upgrade itself.

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -17,15 +17,46 @@ Ostro OS based software can be deployed to a target device in two ways:
-on a CI (Continuous Integration) server and not on the device itself.
+The software update server is a tool used to produce the stream of updates
+that, later on, a device can consume, to upgrade itself.
+It runs on the backend machine(s) producing the build artefacts.
```
The update server runs...

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -35,9 +66,7 @@ and consumed by the software update client running on supported devices.
 
 Update information is secured by generating hashes of each file and then
```
Is this true today?  I've read comments that we're not generating this hash currently.

igor-stoppa replied:
The missing piece is the handling of the signature in the client. The server is signing both individual manifests and the Manifest of Manifests. I've just built an image with an extra bundle and the corresponding update stream and it has all the files required, plus signatures.

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
 
-There is one main os-core bundle containing everything that is mandatory for having
-a bare-bones working system.
+Ostro users coming from popular Linux distros are probably already familiar with
```
Ostro OS users...

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+the concept of package.
+How is a bundle different from a pacakge?
+
+- A bundle doesn't have any dependency, it is self contained.
```
A bundle has no dependencies: it is self-contained. 
A bundle might also list files contained in other bundles, and that is permitted by design.
The software update server and client tools work together to properly handle any such overlapping.

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+  It might list files that are used also by other bundles, but that is by design.
+  Such overlapping is handled by the  Server/Client pair of Software Update tools.
+
+- As long as they come from the same SW Update stream, bundles do not risk to cause
```
As long as bundles come from the same SW update stream, they cannot cause

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+- As long as they come from the same SW Update stream, bundles do not risk to cause
+  installation conflicts on the device, because they were obtained from the same rootfs.
+  It is expected that conflicts were solved during the creation of the rootfs.
+  The person/entity managing the swupd stream is responsible of vetting the rootfs
```
The team managing...

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+- Bundle creation implies that there is a starting rootfs that provides all the files
+  the bundles will include. How this rootfs is created is irrelevant from the
+  persective of the Swupd Server tool, which creates the bundles.
+  In Ostro, the rootfs is populated through the execution of bitbake recipes, but
```
In Ostro OS, ...

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+  In Ostro, the rootfs is populated through the execution of bitbake recipes, but
+  in Clear Linux this is done by installing .rpm packages.
+  Compared to packages, bundles are a competing way to deliver files to the target
+  device, however they can be inserted in an existing workflow that is based on
```
s/they/bundles/

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
+called "os-core", which contains the minimum set of files needed to have a
+barebone working system.
+
 Everything else is optional and can be either deployed or removed by using the
```
s/Everything/All other bundles/

dbkinder commented on doc/architecture/software-update.rst:
```diff
@@ -48,22 +77,55 @@ and applying the updates.
 Everything else is optional and can be either deployed or removed by using the
 software update client program.
 
 Ostro OS developers can also use bundles by following the rule that each application is
```
Ostro OS developers can also use bundles by packaging each application as its own separate bundle.

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+- additional (fast) disk space: the SW Update tool is provided as
+  additional layer, therefore it doesn't require any additional
+  element.
+  However, the process of generating the bundles/updates can be
```
The process of generating the bundles and updates can be very disk intensive and dramatic speed improvements come from using a fast disk storage system, including use of SSD and RAID arrays.

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+  However, the process of generating the bundles/updates can be
+  punishing for the disk.
+  Using a fast disk (ex: SSD/RAID) can bring dramatic improvements.
+  An execution round will also produce various rootfs directory as
```
directories (plura)?

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+  Using a fast disk (ex: SSD/RAID) can bring dramatic improvements.
+  An execution round will also produce various rootfs directory as
+  intermediate artefacts.
+  The actual disk footprint depends on the bundles selected/defined,
```
selected and defined,

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+  simple build, like one might be used to, with a typical distribution
+  based on OECore/YP
+- signing keys (alternatively testing keys from the :file:`swupd-server`
+  project can be used).
```
[Can we reference the certificate-handling document here for more information?  if so, add, "See the :ref:`certificate-handling` technical note for more information about signing keys." ]

pohly replied:
Once we actually have signing keys for swupd, then certificate-handling.rst indees should get its own section about those. But right now we don't have them.

igor-stoppa replied:
Not configurable, but the server is producing signatures for the manifests.
But yes, I agree, once the whole signing is sorted out, the docs need a refresh.

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+==================
+
+The extensive documentation for using SW Update in YP context can be
+found in the `swupd layer`_ and will not be replicated here.
```
[I've seen some requests to include *some* that information in this document.  Is that request dropped for now?]

igor-stoppa replied:
the "some" is what follows, in the next section of the document

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+
+The extensive documentation for using SW Update in YP context can be
+found in the `swupd layer`_ and will not be replicated here.
+These instructions are based on the assumption that the work will happen
```
These bundle creation instructions work with an image that has ``swupd`` enabled.  [Needs a quick explanation of how to get an image with this enabled.]

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+
+- set the `OS_VERSION` variable to an integer value matching `VERSION_ID`
+  in the os-release recipe.
+  This number must be increased before each build that should generate
```
s/that should generate/that generates/

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+These instructions are based on the assumption that the work will happen
+on top of an image flavor that has swupd enabled.
+
+These are the main steps:
```
these steps look like we're making changes to a configuration file -- what file would that be?

igor-stoppa replied:
I can indicate the standard location, but in practice there are so many different ways with bitbake to assign values to parameters, that the user might decide to implement a separate one (for example a custom .inc file)

mythi commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+These are the main steps:
+
+- set the `OS_VERSION` variable to an integer value matching `VERSION_ID`
+  in the os-release recipe.
```
is recipe the right location? AFAICS, it's local.conf or distro conf by default and CI sets it for us.

igor-stoppa replied:
It meant that VERSION_ID is inside os-release.bb
I will reword it.

mythi commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,57 @@
+
+This technical note provides instructions supporting the creation
+of a repository for feeding updates to devices based on Ostro OS.
+
```
My feeling based on discussions with people who don't know swupd is that they may think 'swupd-server' is the web server providing the updates. It is mentioned in this doc that swupd-server is a tool used to create the content but I'm thinking it would make sense to explicitly mention that if the updates are published online a web server is also needed.

igor-stoppa replied:
It was already mentioned explicitly: 
"The information produced by the server is then provided by an HTTP server and consumed by the software update client running on supported devices."

In the new version I've mentioned it twice, also in a note.

Of course 2, 3 or more times doesn't make a difference if the document is not read.

mythi replied:
OK, you can ignore my comment then. I only read the diff this time. Looks OK.

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,61 @@
+Software Update: Building an Ostro |trade| OS  Repository
+#########################################################
+
+This technical note provides instructions supporting the creation
```
This technical note provides instructions for creating a repository used for updating devices based on Ostro OS.

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,61 @@
+=============
+
+- additional (fast) disk space: the SW Update tool is provided as
+  additional layer, therefore it doesn't require any additional
```
an additional layer

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,61 @@
+  disk intensive and dramatic speed improvements can come from
+  using a fast disk storage system, including use of SSD and RAID
+  arrays.
+  An execution round will also produce various rootfs directoriesy
```
s/directoriesy/directories/

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,61 @@
+  arrays.
+  An execution round will also produce various rootfs directoriesy
+  as intermediate artefacts.
+  The actual disk footprint depends on the bundles selected and/or
```
change and/or to just or

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,61 @@
+Bundles definition
+==================
+
+The extensive documentation for using SW Update in YP context can
```
s/YP context/Yocto Project/

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,61 @@
+The extensive documentation for using SW Update in YP context can
+be found in the `swupd layer`_ and will be only summarized here.
+These instructions are based on the assumption that the work will
+happen on top of an image flavor that has swupd enabled .
```
delete the space before the period.

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,61 @@
+These instructions are based on the assumption that the work will
+happen on top of an image flavor that has swupd enabled .
+Example:
+``bitbake ostro-image-swupd`` or download one containing ``-swupd``
```
... or download a pre-compiled image with ``-swupd`` in its name from https:/download.ostroproject.org

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,61 @@
+
+These are the main steps:
+
+- In `build/conf/local.conf` or equivalent location, set the
```
use two back-ticks around filename, ``build/local/local.conf``

dbkinder commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,61 @@
+  swupd update artefacts.
+
+- Assign a list of bundle names to `SWUPD_BUNDLES` i.e:
+
```
You can indicate pre-formatted text (like lines in local.conf) like this:

```
Assign a list of bundle names to ``SWUPD_BUNDLES``
::
   SWUPD_BUNDLES = "feature_one feature_2"

continue with your document
```
(use :: to start the block of text and leave a blank line to indicate the end of the block)

kad commented:
Practicalities for such big reviews: @dbkinder it would be easier if all your changes are sent to @igor-stoppa repository/branch.

dbkinder commented:
OK @kad I just made all my suggested changes as a PR to @igor-stoppa 's branch.

igor-stoppa commented:
@dbkinder are you ok with me squashing your patch? I will of course leave you as co-author, but I would prefer to keep the number of patches at a minimum, related to content changes.

dbkinder commented:
Of course, please squash as needed.
-- david

From: Igor Stoppa [mailto:notifications@github.com]
Sent: Monday, April 25, 2016 1:41 AM
To: ostroproject/meta-ostro <meta-ostro@noreply.github.com>
Cc: Kinder, David B <david.b.kinder@intel.com>; Mention <mention@noreply.github.com>
Subject: Re: [ostroproject/meta-ostro] Software Update Architecture: enhancements. (#130)


@dbkinder<https://github.com/dbkinder> are you ok with me squashing your patch? I will of course leave you as co-author, but I would prefer to keep the number of patches at a minimum, related to content changes.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/ostroproject/meta-ostro/pull/130#issuecomment-214210265>

igor-stoppa commented:
I've added the section "Creating a Custom Image and matching Software Update Repository", the remaining part is basically untouched: the only 2 modifications I did to what was already reviewed are splitting 2 lines, to fit within the 80 columns limit.

I would consider this version of the doc to be final (of course apart from fixing whatever issue might be found).

@pohly @rojkov @gvancuts @dbkinder could you please help reviewing?

gvancuts commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,279 @@
+   file, where to modify the variables that define the content
+   of os-core.
+
+   Example::
```
I don't understand the example, is this a list of variables that should be modified in the recipe/.bbappend to modify the content of os-core? If so, can you make this clearer and perhaps list them out in separate lines.

igor-stoppa replied:
yes, these are variable that must be adjusted, I'll elaborate more on this

gvancuts commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,279 @@
+It is also possible to choose which bundles will be pre-installed on the
+customized image, by:
+
+#. adding the bundle to the list of images generated for bundle purposes
```
I'm not 100% I understand this part. Are we talking about image variants?

Would such variable be tweaked in conf/local.conf?

igor-stoppa replied:
On 26 April 2016 at 14:27, Geoffroy Van Cutsem <notifications@github.com>
wrote:

> I'm not 100% I understand this part. Are we talking about image variants?
>

​This is highly subjective.
From our (Ostro) perspective, they might be thought as image variants​,
however if I have to put myself in the shoes of someone who is making a
distro based on Ostro, it's not a variant at all, it's my base
configuration.

The fact that it doesn't match what Ostro proposes as default, is really
irrelevant, I think.

For the distro builder, the base image is whatever (s)he chooses to be.

> Would such variable be tweaked in conf/local.conf?
>

​I would expect it to possibly start in local.conf, when experimenting, to
eventually move to a .bbappend or possibly a different recipe.

Again, it's hard to predict how the user will use the tools available.
Because of the way bitbake works (it basically slaps together everything in
an humongous single file), many places are equally effective for injecting
a change. There are recommendations available, but it's not so different
from debating (in C) into which of the many included header files to put a
macro : in the end the preprocessor will squash everything into a single
file.
And bitbake is not too different, from that perspective.


-- 
igor

igor-stoppa replied:
@gvancuts did my explanation make this any clearer? In light of it, do you propose any modification?

igor-stoppa replied:
I submitted a new version with improvements for the other point.
If that is sufficiently clear, this is the only issue still pending.

pohly replied:
"adding the bundle to the list of images generated for bundle purposes" is misleading. What gets added in the example is a new bundle image called "pre_installed_content", not a bundle.

gvancuts replied:
Your explanation makes sense @igor-stoppa  but I find that line confusing when you read it (and bear in mind this may be because I don't know enough about this yet... but our users may not know any more initially). I don't understand what the SWUPD_IMAGES variable does and what those "images generated for bundle purposes" are (and how they're used).

igor-stoppa replied:
@pohly: ok
@gvancuts: I will add some explanation of how the bundles are generated to make those points you listed more clear.

gvancuts commented:
Thanks @igor-stoppa , some minor questions left from me but this is looking good!

gvancuts commented on doc/architecture/software-update.rst:
```diff
@@ -17,53 +17,128 @@ Ostro OS based software can be deployed to a target device in two ways:
+called "os-core", which contains the minimum set of files needed to have a
+barebone working system.
+
+All other bundles are optional and can be either deployed or removed by using the
```
Can we actually already remove bundles or is it something that will be implemented later?

igor-stoppa replied:
On 26 April 2016 at 14:34, Geoffroy Van Cutsem <notifications@github.com>
wrote:

>  +All other bundles are optional and can be either deployed or removed by
> using the
>
> Can we actually already remove bundles or is it something that will be
> implemented later?
>

​It works, I confirmed it on the image​

​I cre​ated with the preinstalled bundle and I was able to remove it.

The fact it requires to lookup the remote manifest might not be acceptable
in certain corner cases, but overall I think it's ok.


-- 
igor

gvancuts replied:
I agree, thanks for the clarification

pohly commented on doc/architecture/software-update.rst:
```diff
@@ -17,53 +17,128 @@ Ostro OS based software can be deployed to a target device in two ways:
+
+The Server secures the update information by generating hashes of each
+file (in a manifest) and then signing the file containing the references
+to the hashes (signature of the manifest).
```
That leads to the question of how the signing keys get configured. If that is work in progress, say so here and make it clear that everyone using the current SW Update Server integration is using the same, insecure keys (I believe that is the case?).

Add a note that securing the transport via https is the recommended solution in the meantime.

igor-stoppa replied:
I would prefer to explain it later.
In this case I can replace HTTP with "web" and leave the detailed explanation to the point where the web server is actually needed.

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,281 @@
+- In ``conf/local.conf`` or equivalent location, set the
+  `OS_VERSION` variable to an integer value.
+  This number must be increased before each build that generates
+  swupd update artefacts.
```
This is not required. The automatic default in local builds works well enough to make explicit setting of OS_VERSION an optional step.

igor-stoppa replied:
ok

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,281 @@
+Step 0: Review and optionally modify the content of the os-core bundle
+----------------------------------------------------------------------
+The os-core bundle comes with a predefined set of components,
+however each device is likely to have slighlty different requirements.
```
slightly

igor-stoppa replied:
ok

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,281 @@
+   - IMAGE_INSTALL
+
+   The recommended way is to modify a copy of the original recipe
+   ``ostro-image.bb`` and refer to ``ostro-image.bbclass`` for
```
There is no ostro-image.bb. You probably meant ostro-image-swupd.bb?

igor-stoppa replied:
yes, good catch

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,281 @@
+       pre_installed_content \
+     "
+
+#. adding a varflag to SWUPD_IMAGES.
```
Instead of explaining the syntax, better mention what this does, as in "Define the content of the "pre_installed_content" image by setting the corresponding variable flag in SWUPD_IMAGES:"

igor-stoppa replied:
ok

pohly replied:
You wanted to change "adding a varflag to SWUPD_IMAGES." didn't you?

From the earlier comment: Instead of explaining the syntax, better mention what this does, as in "Define the content of the "pre_installed_content" image by setting the corresponding variable flag in SWUPD_IMAGES:"

igor-stoppa replied:
On 27 April 2016 at 16:07, Patrick Ohly <notifications@github.com> wrote:

> In doc/howtos/software-update-server.rst
> <https://github.com/ostroproject/meta-ostro/pull/130#discussion_r61253118>
> :
>
> > +
> > +
> > +Step 2: Review and optionally modify the content of the images created
> > +----------------------------------------------------------------------
> > +It is also possible to choose which bundles will be pre-installed on the
> > +customized image, by:
> > +
> > +#. adding the bundle to the list of images generated for bundle purposes
> > +
> > +   Example::
> > +
> > +     SWUPD_IMAGES += "\
> > +       pre_installed_content \
> > +     "
> > +
> > +#. adding a varflag to SWUPD_IMAGES.
>
> You wanted to change "adding a varflag to SWUPD_IMAGES." didn't you?
>
> From the earlier comment: Instead of explaining the syntax, better mention
> what this does, as in "Define the content of the "pre_installed_content"
> image by setting the corresponding variable flag in SWUPD_IMAGES:"
>

​I noticed I could improve the previous one, I did it and
 this somehow slipped.​

-- 
igor

igor-stoppa replied:
On 27 April 2016 at 16:22, Stoppa, Igor <igor.stoppa@intel.com> wrote:

> I noticed I could improve the previous one, I did it and
>  this somehow slipped.

Replying from email seems to have hidden the initial comment and my reply.
-- 
igor

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,281 @@
+   syntax. The other identifiers are under the developer's control.
+
+
+This will add to the image content the ``sudo_feature`` feature from Step 1.
```
"sudo_bundle", see earlier comment.

igor-stoppa replied:
What earlier comment?

This one?
"adding the bundle to the list of images generated for bundle purposes" is misleading. What gets added in the example is a new bundle image called "pre_installed_content", not a bundle.

pohly replied:
No, the comment saying that "sudo_feature" should better be renamed to "sudo_bundle". If you rename it, here's another location where "feature" is used twice.

igor-stoppa replied:
This is strange: I do not see such comment.

pohly replied:
I can't find it either, which isn't surprising. Github does not send me a copy of my comments (I already complained to them about that) and comments on old commits also tend to disappear. I'll repeat the comment on the latest commit.

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,281 @@
+
+    $ bitbake ostro-image-swupd-pre_installed_content
+
+Assuming that the chosen architecture was ``intel-corei7-64``, the yield
```
s/yield/output/ ?

igor-stoppa replied:
Initially I used output, but yield seemed more fitting.
yield: to produce (something) as a result of time, effort, or work.

Given the time/resource intensive process required to generate the images, it seemed fitting :-)

Of course I might be wrong, as I'm not a native English speaker.

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,281 @@
+Assuming that the chosen architecture was ``intel-corei7-64``, the yield
+from the command in the Example is:
+
+- The image with the chosen pre-istalled bundles::
```
installed

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,281 @@
+
+    $ bitbake ostro-image-swupd-pre_installed_content
+
+The yield is similar to the previous invocation, however now it also
```
yield again

cschaufler commented:
We have evidence that the lack of checksums over metadata is a problem.The question is whether it's "just" a functional problem or if it's a security problem. If you want this document to meet SDL requirements I think you want to address why the lack of checksums on the version number is not a security problem.

pohly commented:
Igor, I suggest adding an explanation that the existing https support protects sufficiently against online attacks (like man-in-the-middle attacks). The known gap is that offline attacks are possible, which is a known limitation of Ostro OS in general and not specific to the update mechanism.

@cschaufler are you okay with such a description?

igor-stoppa commented:
ok

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,281 @@
+Example::
+
+  SWUPD_BUNDLES += "sudo_feature"
+  BUNDLE_CONTENTS[sudo_feature] = "sudo"
```
I find naming a bundle "feature" misleading, because "feature" is also used in other contexts (image features, distro features) and bundles are (with very few exceptions, see the dev-pkgs feature) explicitly not about those.

I would call this the "sudo_bundle" because it is the bundle which contains sudo. When renaming, also make sure you catch other usages of "feature", not just the actual sudo_feature word.

I understand that this bundle also adds a "feature" to the image (the ability to elevate privileges) but for the sake of this document focusing on "this is a bundle" rather than "this is what the bundle does" is IMHO better.

igor-stoppa commented:
This last version should address all the remaining comments.
Main differences are:

- added notes section about deltas beign created only against version 0
- moved security discussion to own section, with theory and actual status and possible mitigations.

pohly commented on doc/architecture/software-update.rst:
```diff
@@ -17,53 +17,198 @@ Ostro OS based software can be deployed to a target device in two ways:
+in a meta-manifest, called "Manifest of Manifests" or MoM file.
+This file is signed as well.
+
+The design allows for different signign keys, so that the more onerous work of
```
s/signign/signing/

pohly commented on doc/architecture/software-update.rst:
```diff
@@ -17,53 +17,198 @@ Ostro OS based software can be deployed to a target device in two ways:
+The BIOS must also be configured to load exlusively binary files that are signed
+with a known key.
+
+Once the booting is secured, other features (SMACK, IMA) support the extension
```
Smack isn't involved in the chain of trust. I would only mention IMA here.

pohly commented on doc/architecture/software-update.rst:
```diff
@@ -17,53 +17,198 @@ Ostro OS based software can be deployed to a target device in two ways:
+with a known key.
+
+Once the booting is secured, other features (SMACK, IMA) support the extension
+of the chain of trust to the SW Update Client, which is then reponsible of
```
s/reponsiblee/responsible/

pohly commented on doc/architecture/software-update.rst:
```diff
@@ -17,53 +17,198 @@ Ostro OS based software can be deployed to a target device in two ways:
+
+One interested in hardening the security of a self-generated image, must provide
+own keys, apply them to the various signing phases, install them to the BIOS and
+enable EFI-only boot. The update stream will have to flow only through https 
```
Missing . and one extra blank line.

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,285 @@
+.. _software-update-server:
```
Missing URL?

dbkinder replied:
No, this is fine.  It means, define a label here (like an anchor tag in HTML).

pohly commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,285 @@
+
+The process is exemplified by using real components and describing
+what each step looks like, when applied to such components.
+Unless stated otherwise, the changes described are to be intended
```
s/are to be intended/are intended/

pohly commented:
I noticed some more language nitpicks, but overall I'm okay with the proposed change now.

igor-stoppa commented:
One more round ...

barbieri commented on doc/howtos/software-update-server.rst:
```diff
@@ -0,0 +1,287 @@
+
+     SWUPD_IMAGES[pre_installed_content]="\
+       sudo_bundle \
+       sed_bundle \
```
this kind of example will lead to confusion since it is so similar to a "single package", which we want to avoid. Let's try to come with something that encompasses multiple packages and is more like a packagegroup or a task, things like: c-devenv, it would pull editors (vim/emacs), gcc, gdb, valgrind, ...

pohly replied:
@barbieri: can we take that as a suggestion for an improvement after merging the current version?

We need to get this PR reviewed by our security team and then merged as reviewed to cover the security aspects; the rest of the text is unrelated to that, but if we keep revising as part of this PR, we also delay the security part.

barbieri replied:
my suggestion was "c-devenv". Another one can be "powernettools" with wireshark, tcpdump, aircrack... 

we don't need to provide those (and likely can't :-D) so makes a nice case as the actual bundle will never be provided by us.

igor-stoppa replied:
@barbieri I can't stand half done examples - I take it almost personally when I'm reading docs and the author leaves out the hard and actually useful/interesting part :-)  I do not want to do the same :-(
Yes, I agree that you have a point about showing something where bundles do not match 1:1 to packages, but I think the start is already quite complex, without introducing bundles wit multiple packages.
I can add a second example, addressing what you pointed out. And, being the second, I can avoid repeating most of the comments and just focus on the recipe. Only, it needs to be something that the reader can repeat adily, so it must use only components that are whitelisted in Ostro and already working.

barbieri replied:
ok, so we can use something like a second example with more concrete bundle, like add some layer we still do not support and pull in some nice feature :-)

cschaufler commented:
I believe we're good from a security standpoint. +whatevernumberisappropriate.
